doc/designs/thread-api.md

```diff
@@ -0,0 +1,304 @@
+Thread Pool Support
+===================
+
+OpenSSL wishes to support the internal use of threads for purposes of
+concurrency and parallelism in some circumstances. There are various reasons why
+this is desirable:
+
+  - Some algorithms are designed to be run in parallel (Argon2);
+  - Some transports (e.g. QUIC, DTLS) may need to handle timer events
+    independently of application calls to OpenSSL.
+
+To support this end, OpenSSL can manage an internal thread pool. Tasks can be
+scheduled on the internal thread pool.
+
+There are three options given to an application which wants to use the thread
+pool functionality:
```

Comment:
Why do we need three options?
We have to support new APIs ~forever, and should demand strong justifications for variations on things that are already doable with other APIs.  I see some modest argument for a fully automatic default functionality and one other mode, but why are both custom-spawner and manual methods needed?

Reply:
I wouldn't say I necessarily expect all of these to be implemented immediately. They're the reasonable options for an API I see and we probably want to start with just one or two of them.